internal/component/prometheus/remote/queue/network/loop.go

```diff
@@ -0,0 +1,374 @@
+	externalLabels map[string]string
+	series         []*types.TimeSeriesBinary
+	self           actor.Actor
+	ticker         *time.Ticker
+	req            *prompb.WriteRequest
+	buf            *proto.Buffer
+	sendBuffer     []byte
+}
+
+func newLoop(cc types.ConnectionConfig, isMetaData bool, log log.Logger, stats func(s types.NetworkStats)) *loop {
+	// TODO @mattdurham add TLS support afer the initial push.
+	l := &loop{
+		isMeta: isMetaData,
+		// In general we want a healthy queue of items, in this case we want to have 2x our maximum send sized ready.
+		seriesMbx:      actor.NewMailbox[*types.TimeSeriesBinary](actor.OptCapacity(2 * cc.BatchCount)),
+		batchCount:     cc.BatchCount,
+		flushFrequency: cc.FlushFrequency,
+		client:         &http.Client{},
+		cfg:            cc,
+		log:            log,
+		statsFunc:      stats,
+		externalLabels: cc.ExternalLabels,
+		ticker:         time.NewTicker(1 * time.Second),
+		buf:            proto.NewBuffer(nil),
+		sendBuffer:     make([]byte, 0),
+	}
+	l.req = &prompb.WriteRequest{
+		// We know BatchCount is the most we will ever send.
+		Timeseries: make([]prompb.TimeSeries, 0, cc.BatchCount),
+	}
+
+	return l
+}
+
+func (l *loop) Start() {
+	l.self = actor.Combine(l.actors()...).Build()
+	l.self.Start()
+}
+
+func (l *loop) Stop() {
+	l.stopCalled.Store(true)
+	l.self.Stop()
+}
+
+func (l *loop) actors() []actor.Actor {
+	return []actor.Actor{
+		actor.New(l),
+		l.seriesMbx,
+	}
+}
+
+func (l *loop) DoWork(ctx actor.Context) actor.WorkerStatus {
+	// Main select loop
+	select {
+	case <-ctx.Done():
+		l.stopCalled.Store(true)
+		return actor.WorkerEnd
+	// Ticker is to ensure the flush timer is called.
+	case <-l.ticker.C:
+		if len(l.series) == 0 {
+			return actor.WorkerContinue
+		}
+		if time.Since(l.lastSend) > l.flushFrequency {
+			l.trySend(ctx)
+		}
+		return actor.WorkerContinue
+	case series, ok := <-l.seriesMbx.ReceiveC():
+		if !ok {
+			return actor.WorkerEnd
+		}
+		l.series = append(l.series, series)
+		if len(l.series) >= l.batchCount {
+			l.trySend(ctx)
+		}
+		return actor.WorkerContinue
+	}
+}
+
+// trySend is the core functionality for sending data to a endpoint. It will attempt retries as defined in MaxRetryBackoffAttempts.
+func (l *loop) trySend(ctx context.Context) {
+	attempts := 0
+attempt:
+	start := time.Now()
+	result := l.send(ctx, attempts)
+	duration := time.Since(start)
+	l.statsFunc(types.NetworkStats{
+		SendDuration: duration,
+	})
+	if result.successful {
+		l.sendingCleanup()
+		return
+	}
+	if !result.recoverableError {
+		l.sendingCleanup()
+		return
+	}
+	attempts++
+	if attempts > int(l.cfg.MaxRetryBackoffAttempts) && l.cfg.MaxRetryBackoffAttempts > 0 {
+		level.Debug(l.log).Log("msg", "max retry attempts reached", "attempts", attempts)
+		l.sendingCleanup()
+		return
+	}
+	// This helps us short circuit the loop if we are stopping.
+	if l.stopCalled.Load() {
+		return
+	}
+	// Sleep between attempts.
+	time.Sleep(result.retryAfter)
+	goto attempt
```

Comment:
Could we use a `for {}` instead of `goto`? It does look safe here, but we don't know how this software will evolve over time... and I generally resonate with the sentiment against using `goto` as [some raise here](https://www.reddit.com/r/golang/comments/grglzm/why_are_so_many_golang_programmers_against_using/).